invenio_swh/tasks.py

```diff
@@ -74,29 +81,40 @@ def poll_deposit(self, id_):
         StillWaitingException: If the deposit status is "waiting".
 
     """
-    deposit = service.read(id_).deposit
-    service.sync_status(deposit.id)
-    new_status = deposit.status
+    try:
+        deposit = service.read(id_).deposit
+        service.sync_status(deposit.id)
+    except DepositFailed:
+        # If the deposit already failed, we don't need to retry.
+        return
 
     # Manually set status to FAILED on last retry.
     # Celery has a bug where it doesn't raise MaxRetriesExceededError, therefore we need to check retries manually.
     if self.request.retries == 5:
-        service.handle_status_update(deposit, SWHDepositStatus.FAILED)
+        service.update_status(deposit, SWHDepositStatus.FAILED)
         return
 
-    if new_status == SWHDepositStatus.WAITING:
+    if deposit.status == SWHDepositStatus.WAITING:
         raise DepositWaiting("Deposit is still waiting")
 
 
-# @shared_task()
-# def retry_failed_deposits():
-#     try:
-#         result = service.get_unfinished_deposits()
-#         for deposit in result:
-#             res = service.fetch_remote_status(deposit)
-#             status = res.get("deposit_status")
-#             service.update_status(deposit, status)
-#             if res.get("deposit_swh_id"):
-#                 service.update_swhid(deposit, res.get("deposit_swh_id"))
-#     except Exception as exc:
-#         current_app.logger.exception(str(exc))
+@shared_task()
+def cleanup_depositions():
+    """Cleanup old depositions."""
+    if not current_app.config.get("SWH_ENABLED"):
+        current_app.logger.warning()
+        return
+    Deposit = service.record_cls.model_cls
+    # query for records that are stuck in "waiting" and were created in the previous 24 hours
+    res = Deposit.query.filter(
+        Deposit.status == SWHDepositStatus.WAITING,
+        Deposit.created >= datetime.now(timezone.utc) - timedelta(days=1),
+    )
+
+    for deposit in res:
+        try:
+            service.update_status(deposit, SWHDepositStatus.FAILED)
```

Comment:
For reviewers: maybe we can try a last sync with SWH? Perhaps we can get the "FAILED" or "COMPLETED" status from them

Reply:
Yeah, that might be worth actually... in case e.g. they were down and reprocessed their backlog.

We do a similar [trick for the file integrity checks](https://github.com/inveniosoftware/invenio-app-rdm/blob/a7f98e4d6eeed0a5f4d9cca10eaa453a4e55c7f0/invenio_app_rdm/tasks.py#L20), where we retry one last time before generating a report of failures (to make sure they were not transient).